default/tests/shared/database/source_control/test_working_dir.py

```diff
@@ -66,32 +68,102 @@ def test_list(self):
             'access_type': 'restricted'
         }, self.session)
 
+        mock_perm_result = Mock()
+        mock_perm_result.allow_all = False
+        mock_perm_result.allowed_object_id_list = [ds1.id, ds2.id, ds3.id]
+        mock_get_dataset_viewing_permissions.return_value = mock_perm_result
+
         dirs = WorkingDir.list(
             session = self.session,
             project_id = self.project.id,
             member = user.member,
         )
         dirs_id_list = [x.id for x in dirs]
-        self.assertEqual(len(dirs), 6)
+        self.assertEqual(len(dirs), 3)
         self.assertTrue(ds1.id in dirs_id_list)
         self.assertTrue(ds2.id in dirs_id_list)
         self.assertTrue(ds3.id in dirs_id_list)
 
-        RoleMemberObject.new(
-            session = self.session,
-            object_type = ValidObjectTypes.dataset,
-            default_role_name = DatasetDefaultRoles.dataset_viewer,
-            member_id = user.member.id,
-            object_id = ds_restricted.id
-        )
+        mock_perm_result.allowed_object_id_list.append(ds_restricted.id)
+
         dirs = WorkingDir.list(
             session = self.session,
             project_id = self.project.id,
             member = user.member,
         )
         dirs_id_list = [x.id for x in dirs]
-        self.assertEqual(len(dirs), 7)
+        self.assertEqual(len(dirs), 4)
         self.assertTrue(ds1.id in dirs_id_list)
         self.assertTrue(ds2.id in dirs_id_list)
         self.assertTrue(ds3.id in dirs_id_list)
         self.assertTrue(ds_restricted.id in dirs_id_list)
+
+    @patch("shared.database.source_control.working_dir.PolicyEngine")
+    def test_get_dataset_viewing_permissions(self, MockPolicyEngine):
+        # Arrange
```

Comment:
A pattern I commonly use is Arrange, Act, Assert. Arrange is for setup, mocking & defining variables, act is for triggering the action to test, and assert is for assertions.

The idea is to split every test method into these three sections. It helps ensure tests dont get too big, keeps them clean and makes it easy to understand for others. When you're expecting this pattern for a test, its much quicker to parse & understand it. 

Lemme know your thoughts!